### ISSUE TITLE
Java apps don't scale on Haiku's openjdk

### ISSUE THREAD

**Issue by hoper-a11y** [haikuports/haikuports#5136]
They are multithreaded apps and as I saw they also used just one core at a time but not spread the works to all of the cores. This much them much slower when running on Linux where java fully scales to all of the cores. My previous issue about Netbeans crashes also seemed to be related to openjdk's multithreading problems.

I gave the Haiku VM 4 cores and 4096 MB of ram in total. It's the same with both Haiku 32 bit and Haiku 64 bit, java still doesn't scale at all. The openjdk version tested are both openjdk8 and openjdk13.

**Comment by scottmc** [comment]
This should be investigated further to verify if it really is an issue or not.

**Comment by volo-droid** [comment]
I tested it today using a slightly modified version of the following scale test: https://gist.github.com/astromechza/38ab4d097aecf12bcc372575eb01613c . For my tests I increased the numbers of runs from 15 to 300, and tried it both with manually created `Thread` instances, and with `Executors.newFixedThreadPool`. 

I used two setups with either 4 or 6 CPU cores assigned to the Haiku VM (running on VmWare Fusion), and 8 GB of RAM.

Here is my observations:

1. Most of the time the test is scaled as expected
2. Though, **seldomly** two threads are scheduled to run on the same CPU core (I'll call it here _the scaling issue_) resulting in calculation performance degradation
3. The scaling issue can also happen with the setup of 3 or 4 threads, where two of them are assigned to one core
4. When the scaling issue happens, the OS scheduler **usually** is not able to reschedule the running threads to different cores, and the test runs like that till the end
5. **Sometimes** the scheduler can change thread<>CPU assignment over the course of time, fixing the scaling issue. 
6. Disabling/enabling idle CPU cores with the _ProcessController_ applet while running the test **might help** with the rescheduling
7. Disabling/enabling busy CPU core **almost always** fixes the issue
8. Enabling _Power saving_ mode while running the test **usually** cause the scaling issue, but AFAIU that's expected
9. Disabling _Power saving_ mode **might fix** the issue, though usually it doesn't
10. The issue is reproducible within a single JVM, but it's way more easy to reproduce it if several JVMs are launched (i.e. 3 JVMs launched each running the test in 2 threads). 

To sum it up, I think the scaling issue has nothing to do with the JVM itself, and is rather caused by the implementation logic of the OS scheduler.

**Comment by volo-droid** [comment]
|![2 threads](https://user-images.githubusercontent.com/14126089/134772618-ff22a31a-9d00-4c1c-90ca-8e16ea94360a.png)|![3 threads](https://user-images.githubusercontent.com/14126089/134772616-d70c296a-0e47-4b3d-aff4-d573781c3592.png)|
|---|---|
|<img src="https://user-images.githubusercontent.com/14126089/134772482-b8683c19-7d7f-4e18-839a-60a9a2a07ff6.png">|![3 jvms](https://user-images.githubusercontent.com/14126089/134772612-9737a456-94ac-475d-9fdc-2772db8a7077.png)|

**Comment by pulkomandy** [comment]
I don't think there is much use in doing this kind of benchmarks with Haiku running in a virtual machine. Do the problem also happen on real hardware?

**Comment by volo-droid** [comment]
@pulkomandy I have no hardware to try it on, sorry. Why should it matter though? It's not about the performance per se, but the scheduler logic. Frankly speaking, I doubt that would differ on real hardware, but I'd be happy to be proved wrong. If you could test it on 4-core (or more) machine, I'm glad to assist you with the reproducible steps.

**Comment by volo-droid** [comment]
> I don't think there is much use in doing this kind of benchmarks with Haiku running in a virtual machine. Do the problem also happen on real hardware?

I've finally installed Haiku on my laptop, and can confirm that the problem does reproduce on real hardware.
